etcdserver/v3_server.go

```diff
@@ -641,3 +646,89 @@ func (s *EtcdServer) processInternalRaftRequest(ctx context.Context, r pb.Intern
 
 // Watchable returns a watchable interface attached to the etcdserver.
 func (s *EtcdServer) Watchable() mvcc.WatchableKV { return s.KV() }
+
+func (s *EtcdServer) linearizableReadLoop() {
+	var (
+		rs       raft.ReadState
+		requests int
+	)
+	ctx := make([]byte, 8)
+	s.readNotifyc = make(chan struct{})
+
+	for {
+		binary.BigEndian.PutUint64(ctx, s.reqIDGen.Next())
+
+		for (len(s.readwaitc) > 0 && requests < 256) || (len(s.readwaitc) == 0 && requests == 0) {
+			select {
+			case <-s.readwaitc:
+				requests++
+			}
+		}
+
+		s.readMu.Lock()
+		var drained bool
+		for !drained {
+			select {
+			// drain all pending requests
+			case <-s.readwaitc:
+			default:
+				drained = true
+			}
+		}
+		requests = 0
+		notifyc := s.readNotifyc
+		s.readNotifyc = make(chan struct{})
+		s.readMu.Unlock()
+
+		if err := s.r.ReadIndex(context.Background(), ctx); err != nil {
+			if err == raft.ErrStopped {
+				return
+			}
+			plog.Errorf("failed to get read index from raft: %v", err)
+			continue
+		}
+
+		select {
+		case rs = <-s.r.readStateC:
+			if !bytes.Equal(rs.RequestCtx, ctx) {
+				plog.Errorf("unexpected read index context value (want %v, got %v)", rs.RequestCtx, ctx)
+				continue
+			}
+		case <-time.After(time.Second):
+			plog.Warningf("time out waiting for read index response")
+			continue
+		case <-s.done:
+			return
+		}
+
+		if ai := s.getAppliedIndex(); ai < rs.Index {
+			select {
+			case <-s.applyWait.Wait(ai):
+			case <-s.done:
+				return
+			}
+		}
+		// unblock all the l-read that happened before we requested the
+		// read index
+		close(notifyc)
+	}
+}
+
+func (s *EtcdServer) linearizableReadNotify(ctx context.Context) (chan struct{}, error) {
+	for {
+		s.readMu.RLock()
+		select {
+		case s.readwaitc <- struct{}{}:
+			nc := s.readNotifyc
+			s.readMu.RUnlock()
+			return nc, nil
+		case <-ctx.Done():
+			s.readMu.RUnlock()
+			return nil, ctx.Err()
+		default: // chan can be blocking when the read routine is holding the lock for draining
+			s.readMu.RUnlock()
+			// retry after draining
+			time.Sleep(10 * time.Millisecond)
```

Comment:
If we do not access readwaitc and notifyc in one critical section, there might be a race. What can happen is the read routine might wait longer than it should if the run routine finish draining right before the read routine send to s.readwaitc. And the run routine might do unnecessary read request to raft. 

But this dose not really affect correctness I think.